### PR TITLE
OCPP1.6: Improved offline transaction handling

### PR DESCRIPTION
**Pull request by Pietfried**
## Describe your changes
Improved offline transaction handling:
* extended TRANSACTION table in database with START_TRANSACTION_MESSAGE_ID and STOP_TRANSACTION_MESSAGE_ID
* these columns are required in order to process transactionIds received from the CSMS after being offline
* Generating negative transaction ids for transactions to be able to differentiate between multiple offline transactions
* Improved stopping of pending transactions on startup

## Issue ticket number and link
https://github.com/EVerest/libocpp/issues/495

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation
- [x] I read the [contribution documentation](https://github.com/EVerest/EVerest/blob/main/CONTRIBUTING.md) and made sure that my changes meet its requirements

### CODE REVIEW

**Comment by lategoodbye** [comment]
@Pietfried Thanks for the update. Currently i cannot promise that i have the chance to test the fix before 2024.03 is released.

But i like to give some feedback to the implementation. In our propietary stack we also had similiar issues with offline mode. In our transaction database of the OCPP client we decided to store two separate transaction ids (internal & external). At the start of the transaction (regardless on on/offline case) the internal one is generated and stored in the database (external is left empty / NULL). In online case then the StartTransaction.Req is send and in case the StartTransaction.Conf has been received this transaction id is assigned as the external one of the ongoing internal transaction in the database.

This approach has the advantage that the implementation doesn't rely on the assumption that the OCPP CSMS never use negative transaction ids. From our experience at least Virta CSMS use a transaction id = -1in case the start transaction has been rejected (ConcurrentTx).

File: lib/ocpp/v16/charge_point_impl.cpp [hunk 1]
```diff
@@ -781,12 +790,14 @@ void ChargePointImpl::send_meter_value(int32_t connector, MeterValue meter_value
     req.connectorId = connector;
     std::ostringstream oss;
     oss << "Gathering measurands of connector: " << connector;
+
     if (connector > 0) {
         auto transaction = this->transaction_handler->get_transaction(connector);
-        if (transaction != nullptr && transaction->get_transaction_id() != -1) {
-            auto transaction_id = transaction->get_transaction_id();
+        if (transaction != nullptr and transaction->get_transaction_id().has_value()) {
+            // transactions with a positive transaction id have been acknowledged by CSMS
```
Comment (hikinggrass) [inline lib/ocpp/v16/charge_point_impl.cpp:797]: Does this comment still apply?

Comment (Pietfried) [inline lib/ocpp/v16/charge_point_impl.cpp:797]:
```suggestion
```

Comment (Pietfried) [reply lib/ocpp/v16/charge_point_impl.cpp]: Nope :)

**Review by hikinggrass — APPROVED** [review]
Since the database schema changed we should probably add a migration for the next release (or at least warn about this change)

**Pull request merged**